source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc

```diff
@@ -0,0 +1,82 @@
+#include "extensions/compression/brotli/compressor/brotli_compressor_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Compression {
+namespace Brotli {
+namespace Compressor {
+
+BrotliCompressorImpl::BrotliCompressorImpl(const uint32_t quality, const uint32_t window_bits,
+                                           const uint32_t input_block_bits,
+                                           const bool disable_literal_context_modeling,
+                                           const EncoderMode mode, const uint32_t chunk_size)
+    : chunk_size_{chunk_size}, state_(BrotliEncoderCreateInstance(nullptr, nullptr, nullptr),
+                                      &BrotliEncoderDestroyInstance) {
+  RELEASE_ASSERT(quality <= BROTLI_MAX_QUALITY, "");
+  BROTLI_BOOL result = BrotliEncoderSetParameter(state_.get(), BROTLI_PARAM_QUALITY, quality);
+  RELEASE_ASSERT(result == BROTLI_TRUE, "");
+
+  RELEASE_ASSERT(window_bits >= BROTLI_MIN_WINDOW_BITS && window_bits <= BROTLI_MAX_WINDOW_BITS,
+                 "");
+  result = BrotliEncoderSetParameter(state_.get(), BROTLI_PARAM_LGWIN, window_bits);
+  RELEASE_ASSERT(result == BROTLI_TRUE, "");
+
+  RELEASE_ASSERT(input_block_bits >= BROTLI_MIN_INPUT_BLOCK_BITS &&
+                     input_block_bits <= BROTLI_MAX_INPUT_BLOCK_BITS,
+                 "");
+  result = BrotliEncoderSetParameter(state_.get(), BROTLI_PARAM_LGBLOCK, input_block_bits);
+  RELEASE_ASSERT(result == BROTLI_TRUE, "");
+
+  result = BrotliEncoderSetParameter(state_.get(), BROTLI_PARAM_DISABLE_LITERAL_CONTEXT_MODELING,
+                                     disable_literal_context_modeling);
+  RELEASE_ASSERT(result == BROTLI_TRUE, "");
+
+  result = BrotliEncoderSetParameter(state_.get(), BROTLI_PARAM_MODE, static_cast<uint32_t>(mode));
+  RELEASE_ASSERT(result == BROTLI_TRUE, "");
+}
+
+void BrotliCompressorImpl::compress(Buffer::Instance& buffer,
+                                    Envoy::Compression::Compressor::State state) {
+  BrotliContext ctx(chunk_size_);
+
+  for (const Buffer::RawSlice& input_slice : buffer.getRawSlices()) {
```

Comment:
I stole this approach from the gzip encoder since it seems to consume a bit less memory. I guess gzip needs to be updated too.

Reply:
I wonder if it would be better to drain the buffer after every slice, so that the output buffer could potentially re-use the memory from the drained slices from the input buffer.

Reply:
Thanks! Updated to drain the buffer as soon as possible.